pkg/kubelet/kubelet_pods.go

```diff
@@ -291,69 +279,83 @@ func makeMounts(pod *v1.Pod, podDir string, container *v1.Container, hostName, h
 			vol.SELinuxLabeled = true
 			relabelVolume = true
```

Comment:
Do we want to relabel the image volume when the SELinux is enabled on the node?

```go
func (o *imagePlugin) GetAttributes() volume.Attributes {
	return volume.Attributes{
		ReadOnly:       true,
		Managed:        true,
		SELinuxRelabel: true,
	}
}
```

BTW, the image volume is shared among all the pods on the node if the pod's volume is an image. I'm unfamiliar with SELinux, what's the impact if we relabel the image volume?

When SELinux is enabled, can pod1 and pod2 access the same image volume successfully even if Managed is false?

Reply:
I can not enable selinux on kind node. I'm not sure whether the kind does not support selinux.

What's the expected behavior on image volume when SELinux is enabled on node? @saschagrunert

Reply:
Let's pass that information to the runtime and let them decide. For CRI-O we mount directly using the correct mount label, so the indicator should not change anything. 

> When SELinux is enabled, can pod1 and pod2 access the same image volume successfully even if Managed is false

Not when they use different labels, which is a known limitation for volume plugins.

Reply:
Does the pod1 and pod2 have the same mount label (selinux)? 

I see that the image volume has already be labeled when the first pod is created.

https://github.com/cri-o/cri-o/blob/485358b6ef0b69b0a345b3d4e5c62211477aeb8a/server/container_create_linux.go#L1078

Reply:
> Does the pod1 and pod2 have the same mount label (selinux)?

Depends on the platform, but does that matter? We have to fail pod2 if pod1 is already using a different label.

Reply:
> Not when they use different labels, which is a known limitation for volume plugins.

Does the kubernetes website include the known limitation? If not, it is better to add it. Can you update the website to include the known limitation? I'm worried that I can't explain it clearly.

Reply:
Yes, I can take care of that.

Reply:
Closing the loop: https://github.com/kubernetes/website/pull/47588

Reply:
> Yes, I can take care of that.

Thank you.

> Let's pass that information to the runtime and let them decide.

Do we need to set Managed and SELinuxRelabel to false in `GetAttributes` of the image plugin?

the hostPath plugin set these values to false

```go
func (b *hostPathMounter) GetAttributes() volume.Attributes {
	return volume.Attributes{
		ReadOnly:       b.readOnly,
		Managed:        false,
		SELinuxRelabel: false,
	}
}
```

Reply:
I'm not sure honestly, my instinct would say set them to `true` (as-is).

Reply:
> Closing the loop: https://github.com/kubernetes/website/pull/47588

/cc @kerthcet 

When scheduling pods, does the scheduler need to add a new plug-in to adapt to this situation? To avoid the pod from not being able to start up due to SELinux being enabled after the pod is scheduled to the node

/cc @wzshiming 

Does the containerd has the same limitation like cri-o?

Reply:
> I'm not sure honestly, my instinct would say set them to true (as-is).

cc @jsafrane @gnufied can you have a look?

Reply:
@saschagrunert It seems that we can not relabel the volume when the selinux is enabled.

```
              waiting:
                message: 'relabel failed /var/lib/containers/storage/overlay/82099f56ba7822d8408ac8517b90bde2aa036ac32ee2dd3db810348302edd618/merged:
                  lsetxattr /var/lib/containers/storage/overlay/82099f56ba7822d8408ac8517b90bde2aa036ac32ee2dd3db810348302edd618/merged/dir:
                  read-only file system'
                reason: CreateContainerError
```

Reply:
We probably have to considering the test failures in `pull-kubernetes-node-crio-cgrpv2-imagevolume-`e2e

Reply:
Is it possible for image volume to make it act like emptyDir in the cri-o? 

One pod/containers can isolate the volume from another pod/containers. Don't share the volume between different pods/containers. They can have different mountpoints for the same image.

Reply:
If we can do that, no other requirement will be needed in kube-scheduler. Users can use it more easily without extra configuration.

Reply:
@carlory we need to extend the c/storage API to be able to specify a different mountpoint.

Reply:
But I'm not sure if we can even run into that case, see: https://github.com/containers/storage/issues/1994

Reply:
How about this? @saschagrunert  

1. Extend the `kuberuntime.ImageService` proto to support mount/unmount image: (I'm not sure whether it can be done)

```
// ImageService defines the public APIs for managing images.
service ImageService {
    // MountImage mount the image to the given path.
    rpc MountImage(MountImageRequest) returns (MountImageResponse) {}
    // UnmountImage unmount the image from the given path.
    rpc UnmountImage(UnmountImageRequest) returns (UnmountImageResponse) {}
}
```

2. Image volume plugin:

```go
type imageVolume struct {
	volName string
	podUID  types.UID
	plugin  *imagePlugin
	volume.MetricsProvider
}

var _ volume.Volume = &imageVolume{} 

// allow volume plugin to manage the volume path.
func (iv *imageVolume) GetPath() string { 
	return iv.plugin.host.GetPodVolumeDir(iv.podUID, utilstrings.EscapeQualifiedName(pluginName), iv.volName)
}

type imageVolumeMounter struct {
	*imageVolume

	source v1.ImageVolumeSource
	pod    v1.Pod

	pullImage    func(pod v1.Pod, source v1.ImageVolumeSource) error
	mountImage   func(path string, image string) error // call cri to mount image
	unmountImage func(path string) error // call cri to unmount image
}

var _ volume.Mounter = &imageVolumeMounter{}

func (m *imageVolumeMounter) RequiresFSResize() bool { return false }

func (m *imageVolumeMounter) RequiresRemount(spec *volume.Spec) bool { return false }

func (m *imageVolumeMounter) GetAttributes() volume.Attributes {
	return volume.Attributes{
		ReadOnly:       true,
		Managed:        true,
		SELinuxRelabel: true,
	}
}

func (m *imageVolumeMounter) SetUp(mounterArgs volume.MounterArgs) error {
	return m.SetUpAt(m.GetPath(), mounterArgs)
}

func (m *imageVolumeMounter) SetUpAt(dir string, mounterArgs volume.MounterArgs) error {
	if err := m.pullImage(m.pod, m.source); err != nil {
		return err
	}
	return m.mountImage(dir, m.source.Reference)
}

func (m *imageVolumeMounter) SupportsMountOption() bool { return false }

func (m *imageVolumeMounter) SupportsSELinuxContextMount(spec *volume.Spec) (bool, error) {
	return false, nil
}

func (m *imageVolumeMounter) TearDown() error {
	return m.TearDownAt(m.GetPath())
}

func (m *imageVolumeMounter) TearDownAt(string) error {
	return m.unmountImage(m.GetPath())
}
```
If we can do, 

- maybe we able to support subpath in the future?
- allow more than one pods to running on the same node when using the same image.
- remove specific logic in kubelet pod loop and `makemounts`.

Reply:
@carlory I think changes like this needs to go into a graduation of the feature and are out of scope of this fix here. I appreciate your input here :pray:

Reply:
It's indeed out of scope of this fix. I will mark the above conversation as resolved. 

And if there are discussions on this topic elsewhere in the future, I would be happy to participate if possible.  😊